src/engine/GroupBy.cpp

```diff
@@ -0,0 +1,932 @@
+// Copyright 2018, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Florian Kramer (florian.kramer@mail.uni-freiburg.de)
+
+#include "GroupBy.h"
+
+#include "../index/Index.h"
+#include "../util/Conversions.h"
+#include "../util/HashSet.h"
+
+GroupBy::GroupBy(QueryExecutionContext* qec,
+                 std::shared_ptr<QueryExecutionTree> subtree,
+                 const vector<string>& groupByVariables,
+                 const std::vector<ParsedQuery::Alias>& aliases)
+    : Operation(qec), _subtree(subtree), _groupByVariables(groupByVariables) {
+  _aliases.reserve(aliases.size());
+  for (const ParsedQuery::Alias& a : aliases) {
+    if (a._isAggregate) {
+      _aliases.push_back(a);
+    }
+  }
+  std::sort(_aliases.begin(), _aliases.end(),
+            [](const ParsedQuery::Alias& a1, const ParsedQuery::Alias& a2) {
+              return a1._outVarName < a2._outVarName;
+            });
+
+  // sort the groupByVariables to ensure the cache key is order invariant
+  std::sort(_groupByVariables.begin(), _groupByVariables.end());
+
+  // The returned columns are all groupByVariables followed by aggregrates
+  size_t colIndex = 0;
+  for (std::string var : _groupByVariables) {
+    _varColMap[var] = colIndex;
+    colIndex++;
+  }
+  for (const ParsedQuery::Alias& a : _aliases) {
+    _varColMap[a._outVarName] = colIndex;
+    colIndex++;
+  }
+}
+
+string GroupBy::asString(size_t indent) const {
+  std::ostringstream os;
+  for (size_t i = 0; i < indent; ++i) {
+    os << " ";
+  }
+  os << "GROUP_BY" << std::endl;
+  for (const std::string var : _groupByVariables) {
+    os << var << ", ";
+  }
+  for (auto p : _aliases) {
+    os << p._function << ", ";
+  }
+  os << std::endl;
+  os << _subtree->asString(indent);
+  return os.str();
+}
+
+size_t GroupBy::getResultWidth() const { return _varColMap.size(); }
+
+size_t GroupBy::resultSortedOn() const { return -1; }
+
+vector<pair<size_t, bool>> GroupBy::computeSortColumns(
+    std::shared_ptr<QueryExecutionTree> subtree,
+    const vector<string>& groupByVariables,
+    const std::vector<ParsedQuery::Alias>& aliases) {
+  // Create sorted lists of the aliases and the group by variables to determine
+  // the output column order, on which the sorting depends. Then populate
+  // the vector of columns which should be sorted by using the subtrees
+  // variable column map.
+  vector<pair<size_t, bool>> cols;
+  if (groupByVariables.empty()) {
+    // the entire input is a single group, no sorting needs to be done
+    return cols;
+  }
+
+  std::vector<ParsedQuery::Alias> sortedAliases;
+  sortedAliases.reserve(aliases.size());
+  for (const ParsedQuery::Alias& a : aliases) {
+    if (a._isAggregate) {
```

Comment:
This looks very similar to the code in the constructor, maybe that should call this static function? Also are we sure we aren't sorting the `_alias` thing twice?

Reply:
The code is almost the same. The problem is, that to create the OrderBy operation required by the GroupBy the input columns of group by need to be computet known, but the constructor of an operation normally takes the root operation, so the group by needs to be created before the OrderBy.
A less redundant solution would be, to add a function for setting the parent operation of the group by. That way the column order needs to be computed only once, but GroupBy would be the only operation that is created that way.